Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DataSynchronizer.swift

```diff
@@ -0,0 +1,557 @@
+     - parameter namespace: the namespace to conduct this op
+     - returns: the result of the remove many operation
+     */
+    func deleteMany(filter: Document,
+                    in namespace: MongoNamespace) -> DeleteResult? {
+        fatalError("\(#function) not implemented")
+    }
+
+    /**
+     Update a single document in the collection that have been synchronized with the remote
+     according to the specified arguments. If the update results in an upsert,
+     the newly upserted document will automatically become synchronized.
+
+     - parameter filter: a document describing the query filter, which may not be null.
+     - parameter update: a document describing the update, which may not be null. The update to
+     apply must include only update operators.
+     - parameter namespace: the namespace to conduct this op
+     - returns: the result of the update one operation
+     */
+    func updateOne(filter: Document,
+                   update: Document,
+                   options: UpdateOptions?,
+                   in namespace: MongoNamespace) -> UpdateResult? {
+        fatalError("\(#function) not implemented")
+    }
+
+    /**
+     Update all documents in the collection that have been synchronized with the remote
+     according to the specified arguments. If the update results in an upsert,
+     the newly upserted document will automatically become synchronized.
+
+     - parameter filter: a document describing the query filter, which may not be null.
+     - parameter update: a document describing the update, which may not be null. The update to
+     apply must include only update operators.
+     - parameter updateOptions: the options to apply to the update operation
+     - parameter namespace: the namespace to conduct this op
+     - returns: the result of the update many operation
+     */
+    func updateMany(filter: Document,
+                    update: Document,
+                    options: UpdateOptions?,
+                    in namespace: MongoNamespace) -> UpdateResult? {
+        fatalError("\(#function) not implemented")
+    }
+
+    /**
+     Emits a change event for the given document id.
+
+     - parameter documentId: the document that has a change event for it.
+     - parameter event:      the change event.
+     */
+    private func emitEvent(documentId: BSONValue, event: ChangeEvent<Document>) {
+        listenersLock.readLock()
+        defer { listenersLock.unlock() }
+
+        let nsConfig = syncConfig[event.ns]
+
+        eventDispatchQueue.async {
+            nsConfig?.changeEventListener?.onEvent(documentId: documentId,
+                                                   event: event)
+        }
+    }
+
+    /// Potentially pass along useful error information to the user.
+    /// This should only be used for low level errors.
+    func on(error: Error, for documentId: BSONValue?, in namespace: MongoNamespace?) {
```

Comment:
ah, I'm now realizing that FatalErrorListener as you implemented it is purely internal. This sort of negates my other comments. I don't think we should block on this stuff, so I think this is all fine for now, but let's keep this in mind for the merge into master.